[Enterprise Search] Add `.connector-secrets` system index and GET/POST requests

navarone-feekery opened the pull request:
## Context

The Search team is currently implementing BYOI (Bring Your Own Index) for Connectors. This feature will allow users to create search indices for Connectors without having `search-` prefix the index name.
Native Connectors are run using the Enterprise Search system account, which has access to search indices because it can access all indices that follow the pattern `search-*`. To continue allowing Native Connectors to function after removing the prefix requirement, we need to switch to using API keys to manage index access. That brings us to this PR.

The goal of this PR is to have a place where Native Connectors can store API keys (secrets storage). A connector secret can be stored from either Kibana or the Connectors CLI. Kibana would only have write access to this index so it cannot expose secrets in the UI. Enterprise Search (which runs the Connectors CLI on cloud) will have both read and write access.

These changes follow the implementation design of Fleet's secrets storage ([example PR](https://github.com/elastic/elasticsearch/pull/97728)). The implementation is largely identical, with a change in namespace.

## Changes

- Introduce new internal system index called `.connector-secrets`. This will be used to store the API keys that Native Connectors will use.
- Add GET and POST requests for connector secrets
- Give the Kibana system account write access to `.connector-secrets`
- Give the Enterprise Search account read and write access to `.connector-secrets`

### Not included

- To reduce PR size, DELETE requests will be added in a further PR

### How to test this PR

The API endpoints can be tested in Kibana Dev Tools:

1. Run ES from source with `./gradlew run`.
2. Once ES is running, change the default password for the elastic user from "password" to "changeme" to match Kibana's:
   ```
   curl -k -u elastic-admin:elastic-password -H 'Content-Type: application/json' \
   http://localhost:9200/_security/user/elastic/_password -d'{"password": "changeme"}'
   curl -k -u elastic:changeme -H 'Content-Type: application/json' \
    http://localhost:9200/_security/user/kibana_system/_password -d'{"password": "changeme"}'
   ```
3. Run Kibana with `yarn start`.
4. Log into Kibana and go to `/app/dev_tools#/console`.
5. Create a secret:
   ```
   POST /_connector/secret/
   {
     "value": "test123"
   }
   ```
6. This should succeed and return the id of the created secret doc, e.g.:
   ```
   {
     "id": "1234"
   }
   ```
7. Copy the id and try to get the secret:
   ```
   GET /_connector/secret/1234
   ```
8. This should succeed and return the secret id and value:
   ```
   {
     "id": "1234",
     "value": "test123"
   }
   ```

9. Try to get a different secret using a random value. It should fail with 404 and return an error.
    ```
    GET /_connector/secret/notrealid
    ```

navarone-feekery committed:
Add search secrets index and GET requests

navarone-feekery committed:
Rename to Connector Secrets

navarone-feekery committed:
Fix inconsistent naming

navarone-feekery committed:
Rename cluster privilege

navarone-feekery committed:
Add GET tests

navarone-feekery committed:
Add POST request and tests

navarone-feekery committed:
Add missing tests

navarone-feekery committed:
Fix cluster privilege name

navarone-feekery committed:
Fix more tests

navarone-feekery committed:
Fix cluster privilege tests and add index permissions

navarone-feekery committed:
Add java rest test for system index

navarone-feekery committed:
Fix cluster privilege builtin count

navarone-feekery committed:
Fix cluster privilege test

navarone-feekery committed:
rebuild

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:Enterprise Search)

mark-vieira commented on x-pack/plugin/ent-search/build.gradle:
```diff
@@ -38,6 +38,13 @@ dependencies {
 }
 
+testClusters.configureEach {
+  testDistribution = 'DEFAULT'
```
@breskeby we should think about making it simpler to setup clusters that require security w/o having to use the default distribution. I suspect the vast majority of tests that use the default distro is for this reason alone.

Otherwise these changes LGTM.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/SecretsIndexService.java:
```diff
@@ -0,0 +1,59 @@
+            .setPrimaryIndex(CONNECTOR_SECRETS_INDEX_NAME + "-" + CURRENT_INDEX_VERSION)
+            .setDescription("Secret values managed by Connectors")
+            .setMappings(request.mappings())
+            .setSettings(request.settings())
```
Question, do we want specific settings here? For example, some of our other enterprise search system indices set the number of shards/replicas, and refresh interval. 

Note: If you decide to set a refresh interval below what serverless supports you'll need to update serverless to use fast refresh.

navarone-feekery replied:
I think explicitly defining settings is probably a good thing. I've added the default that the other system indices use.
Regarding fast refresh -- how would I determine if this index requires < 5s refresh interval? I'm not sure what the repercussions are here in either direction.

navarone-feekery replied:
This came up in another comment as well; it appears Fleet uses fast refresh for their secrets storage so that's a good precedent for us to copy them.
https://github.com/elastic/elasticsearch-serverless/blob/a40125757d2c60fc8b9b352d3e235d54cb362d9d/modules/serverless-fleet/src/main/java/co/elastic/elasticsearch/serverless/fleet/ServerlessFleetPlugin.java#L38

I'll open a followup PR to make that change. Thanks!

kderusso commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/KibanaOwnedReservedRoleDescriptors.java:
```diff
@@ -378,7 +379,13 @@ static RoleDescriptor kibanaSystem(String name) {
+                // Connector secrets. Kibana can only write to this index.
+                RoleDescriptor.IndicesPrivileges.builder()
+                    .indices(".connector-secrets*")
+                    .privileges("write", "delete", "create_index")
```
Do we need `read` here as well?

navarone-feekery replied:
Kibana won't have read access. That way the secrets can't be surfaced in the UI.
Originally I wanted Kibana to have write only, and Enterprise Search to have read only. But because Enterprise Search has `manage`  I can't scope it the way I want without massive changes, so I decided having write for Enterprise Search as well is okay.
However I think Kibana shouldn't have read access.

navarone-feekery replied:
I see I had a typo in my PR about read/write access for Kibana, sorry if that caused confusion

kderusso replied:
That makes sense, thanks!

kderusso reviewed (commented):
This looks great! A couple of questions and notes. 

The only other thing that I'd request before merging is adding unit tests extending `AbstractBWCSerializationTestCase` for your request and response objects. After that I'm ready to 👍. Thank you!

timgrein reviewed (commented):
Really nice work and congrats on your (I think) first PR in ES 👏  And thank you for the detailed PR description!

Left some minor comments, which are non-blocking and having one question: Do you plan to add the [Backwards compatibility tests](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/syncjob/action/CancelConnectorSyncJobActionRequestBWCSerializingTests.java), [index service tests](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java) and other unit tests, too?

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/PostSecretRequest.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public ActionRequestValidationException validate() {
+        if (this.value == null) {
```
would it be cleaner to use `Strings.isNullOrEmpty(this.value)` here? Do we allow the value to be empty string?

jedrazb reviewed (commented):
Great work! 🚀  Left some minor comments

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/RestPostSecretAction.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(RestRequest.Method.POST, "/_connector/secret"));
```
Noticed during testing: Should this be `_secret` with a leading underscore for consistency purposes? I think we're always using `_` at other places (`_connector/_sync_job`, `_search` etc.)

jedrazb replied:
++

timgrein reviewed (commented):
One additional question around endpoint naming 🙏

navarone-feekery commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/RestGetSecretAction.java:
```diff
@@ -0,0 +1,38 @@
+import java.io.IOException;
+import java.util.List;
+
+@ServerlessScope(Scope.INTERNAL)
```
Do you know what the repercussions in either direction for this would be? I'm not sure if this index requires a < 5s refresh or not.

jedrazb replied:
Best bet would be to check how fleet handles it in serverless. Refresh rate refers to how quickly your data is searchable in the index.

navarone-feekery replied:
It appears that they've allowed fast refresh so I think that's good precedent for us to do the same: https://github.com/elastic/elasticsearch-serverless/blob/a40125757d2c60fc8b9b352d3e235d54cb362d9d/modules/serverless-fleet/src/main/java/co/elastic/elasticsearch/serverless/fleet/ServerlessFleetPlugin.java#L38

I'll open a PR after to follow through on this. Thanks!

navarone-feekery committed:
Changes based on code review

navarone-feekery committed:
Placate linter

carlosdelest commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccounts.java:
```diff
@@ -49,6 +49,12 @@ final class ElasticServiceAccounts {
                     .privileges("manage", "read", "write")
+                    .build(),
+                // Connector secrets. Enterprise Search can read and write to this index.
+                RoleDescriptor.IndicesPrivileges.builder()
```
I'm confused about the privileges management. What we used in the past is a bit different:
- Define new privileges (like `WRITE_CONNECTOR_SECRETS` and `READ_CONNECTORS_SECRETS`). These privileges are used to restrict actions for users based on the defined patterns (like [here](https://github.com/elastic/elasticsearch/blob/770671667eaad7eca8404785c0422bb54a4213f5/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java#L157)).
- Assign these privileges to the service accounts (or the users that need them), like [here](https://github.com/elastic/elasticsearch/blob/0fca9c3098d446b865ab5071f9b0f29982c32a5f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccounts.java#L65).
- Create internal users for managing the system indices. It's there that the index-level privileges over the system index are defined, like [here](https://github.com/elastic/elasticsearch/blob/c390d666628c425a6355a3e3fc58cfc0fa785d3c/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L164).

Does that make sense? Or is this a conscious decision and I'm missing the point?

carlosdelest replied:
>  There was no point adding one for read/write for Enterprise Search because it has manage as a cluster privilege which gives it full access to the cluster already.

AFAIU `manage` does not provide actual index read permission for the index, even if it provides the tools for eventually having that.

> There shouldn't be a situation (at least in 8.13) where a non-Enterprise Search user needs read access to secrets. That's why I've not included a cluster privilege for it.

If we have created a cluster privilege for Kibana write, why not create a cluster privilege for reading? That seems odd to me.

> When are the internal users "used" in practice?

They are converted to system privileges. Connectors use the internal XPack user [here](https://github.com/elastic/elasticsearch/blob/c390d666628c425a6355a3e3fc58cfc0fa785d3c/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L155), which [has full access to indices](https://github.com/elastic/elasticsearch/blob/770671667eaad7eca8404785c0422bb54a4213f5/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java#L98) IIUC.

This means that using `CONNECTORS_ORIGIN` for a client internally translates it to the `XPACK_USER`, which has full access to system indices including security.

The problem with the current approach is that any user that has access to the get secret action will execute the action using the `OriginSettingClient`, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.

The way to change that would be to either:
- Define a read privilege that would limit access to the get secrets action, or
- Do not use the `OriginSettingClient`, but use the actual user that makes the request and rely on their index permissions for security.

I find defining privileges is better and more explicit than relying on system indices permissions.

I hope this makes sense. If I have misunderstood the security mechanism please call me out - but I would do a quick check with a non-privileged user just to make sure that accessing the secrets is not possible.

ioanatia replied:
Chiming in on this thread, I know that for fleet secrets we have a `write_fleet_secrets` method and that's the pattern we reuse here. If we look at the other built in cluster privileges, the pattern we usually have is to have a `manage_[feature]` privilege paired with a `read_[feature]` privilege:

```
GET _security/privilege/_builtin

response:
{
  "cluster": [
    "all",
    "cancel_task",
    "create_snapshot",
    "cross_cluster_replication",
    "cross_cluster_search",
    "delegate_pki",
    "grant_api_key",
    "manage",
    "manage_api_key",
    "manage_autoscaling",
    "manage_behavioral_analytics",
    "manage_ccr",
    "manage_data_frame_transforms",
    "manage_enrich",
    "manage_ilm",
    "manage_index_templates",
    "manage_ingest_pipelines",
    "manage_logstash_pipelines",
    "manage_ml",
    "manage_oidc",
    "manage_own_api_key",
    "manage_pipeline",
    "manage_rollup",
    "manage_saml",
    "manage_search_application",
    "manage_search_query_rules",
    "manage_search_synonyms",
    "manage_security",
    "manage_service_account",
    "manage_slm",
    "manage_token",
    "manage_transform",
    "manage_user_profile",
    "manage_watcher",
    "monitor",
    "monitor_data_frame_transforms",
    "monitor_enrich",
    "monitor_ml",
    "monitor_rollup",
    "monitor_snapshot",
    "monitor_text_structure",
    "monitor_transform",
    "monitor_watcher",
    "none",
    "post_behavioral_analytics_event",
    "read_ccr",
    "read_fleet_secrets",
    "read_ilm",
    "read_pipeline",
    "read_security",
    "read_slm",
    "transport_client",
    "write_fleet_secrets"
  ],
  "index": [
    "all",
    "auto_configure",
    "create",
    "create_doc",
    "create_index",
    "cross_cluster_replication",
    "cross_cluster_replication_internal",
    "delete",
    "delete_index",
    "index",
    "maintenance",
    "manage",
    "manage_data_stream_lifecycle",
    "manage_follow_index",
    "manage_ilm",
    "manage_leader_index",
    "monitor",
    "none",
    "read",
    "read_cross_cluster",
    "view_index_metadata",
    "write"
  ]
}
```

Maybe we should do the same here - have `manage_connectors_secrets` and `read_connectors_secrets`.
`manage_connectors_secrets` allows for both reads and writes. 
I think this is a better pattern and I am thinking how App Search also has a read/write pattern for API keys. In retrospect I'd wish we went for the same pattern with manage/read.

ioanatia replied:
> The problem with the current approach is that any user that has access to the get secret action will execute the action using the OriginSettingClient, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.

We should have a test for this use case.

navarone-feekery replied:
@carlosdelest @ioanatia I've added the `read_connector_secrets` cluster privilege back in [this commit](https://github.com/elastic/elasticsearch/pull/103683/commits/9cf3f81a9e115dd26c15e24124d087a944ea4543). I've opted not to use `manage_` because this is a little different from our other cluster privileges. I want to have `write_` as a separate privilege because some roles should have write without read permissions.

If we were to use `manage_`, the two cluster privileges would be `write_connector_secrets` and `manage_connector_secrets`. I'm okay with that but I wasn't sure if it aligned with what yous were expecting. I'm happy to change it to `manage_` and `write_`, or keep `read_` and `write_`. WDYT?

> The problem with the current approach is that any user that has access to the get secret action will execute the action using the OriginSettingClient, which internally translates to the XPack user, who has access to the system index. As there is no privilege for reading secrets, any user that access the endpoint will end up being able to read secrets. That's a security issue.
>> We should have a test for this use case.

Do you know of any example tests I can use for reference? I'm not sure what I should even be looking for here.

navarone-feekery replied:
AH sorry, I see you've linked examples in [this comment](https://github.com/elastic/elasticsearch/pull/103683#discussion_r1447101822). I'll try to add some tests for this.

ioanatia replied:
okay, I tested this locally with your branch, it seems we do require `manage` or `all` which I think is on purpose since we don't want to risk adding APIs that don't require a privilege:

```
# with no roles
 curl -XGET "http://localhost:9200/_connector/_secret/-R7z94wBGenV8ZXMozwE" -H "kbn-xsrf: reporting" -u ioana:changeme
{"error":{"root_cause":[{"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [], this action is granted by the cluster privileges [manage,all]"}],"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [], this action is granted by the cluster privileges [manage,all]"},"status":403}%
# with an unrelated role
curl -XGET "http://localhost:9200/_connector/_secret/-R7z94wBGenV8ZXMozwE" -H "kbn-xsrf: reporting" -u ioana:changeme
{"error":{"root_cause":[{"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [machine_learning_user], this action is granted by the cluster privileges [manage,all]"}],"type":"security_exception","reason":"action [cluster:admin/xpack/connector/secret/get] is unauthorized for user [ioana] with effective roles [machine_learning_user], this action is granted by the cluster privileges [manage,all]"},"status":403}
```

It would still be good to add an explicit test that a user without `manage` or `all` cannot access the API, just to catch any regressions.

>  > manage_connectors_secrets allows for both reads and writes.

> I want the user/Kibana to only have write access however. If we use manage_ is it okay to keep it restricted to write only?

Then we would just change the name, which doesn't do much IMO. I don't have enough context to understand why we would want to restrict the user/Kibana to only writes.
For the purpose of this PR I would say:
- let's use `manage` and `all` for writes/deletes
- add tests to check that a user without any of these privileges cannot do writes/deletes on connectors secrets.

Then in a separate PR we can see what we can do about an additional privilege like `write_connectors_secret` /`read_connectors_secret`. WDYT?

navarone-feekery replied:
@ioanatia @carlosdelest I've added tests and fixed up some of the cluster privilege stuff in [this commit](https://github.com/elastic/elasticsearch/pull/103683/commits/54875c354496fdb9a0d2519a32c08ad7e3f2283e).

@ioanatia I'm not sure how much of your comment still applies as we discussed this PR on slack since then. Do you still feel strongly about using `manage` and `all` for writes/deletes? Or have we now covered the issues you mentioned?

carlosdelest replied:
Thanks for iterating on this @navarone-feekery ! 💯

navarone-feekery commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:
```diff
@@ -325,6 +325,11 @@ public class ClusterPrivilegeResolver {
         CROSS_CLUSTER_REPLICATION_PATTERN
     );
 
+    public static final NamedClusterPrivilege WRITE_CONNECTOR_SECRETS = new ActionClusterPrivilege(
```
There was, but I removed it. Originally the `read_connector_secrets` cluster privilege was being applied to the Enterprise Search service account, but because it already has `manage` it can access everything. I can't effectively limit its scope without reworking that.
So in any case, creating a cluster privilege for reads isn't necessary yet.

navarone-feekery replied:
I've re-added the `read_connector_secrets` cluster privilege. A similar conversation is taking place [here](https://github.com/elastic/elasticsearch/pull/103683#discussion_r1447266495) so I'll resolve this one.

navarone-feekery commented on rest-api-spec/src/main/resources/rest-api-spec/api/connector_secret.get.json:
```diff
@@ -0,0 +1,28 @@
+      "description": "Retrieves a secret stored by Connectors."
+    },
+    "stability": "experimental",
+    "visibility":"private",
```
If we set this to `feature_flag` does it make it visible to users? My intention for this endpoint is that it's internal.

carlosdelest replied:
good point, I think it makes sense to keep it private.

navarone-feekery commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/TransportPostSecretAction.java:
```diff
@@ -0,0 +1,45 @@
+
+    protected void doExecute(Task task, PostSecretRequest request, ActionListener<PostSecretResponse> listener) {
+        try {
+            client.prepareIndex(CONNECTOR_SECRETS_INDEX_NAME)
```
Sorry, I don't understand exactly what you mean. Can you expand on this a bit?

carlosdelest replied:
Instead of doing the actual index operation here, I would defer to an operation created in the `SecretIndexService` for that. That is the class that should deal with all operations related to storage and retrieval of the secrets (like a DAO pattern) and understands the internals of the index.

timgrein replied:
++ here are two examples from @jedrazb and me: [ConnectorIndexService](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java) and [ConnectorSyncJobIndexService](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java), hope this helps 🙏

navarone-feekery replied:
Thanks for the clarifications! I've attempted to do this in this commit: https://github.com/elastic/elasticsearch/pull/103683/commits/34c9298711e1b205d353c1aa82d88dd1fc5b55b5

Is this what you're talking about?

timgrein replied:
Yes, exactly, thanks for the change! Small nit: IMHO `createSecret` is a better name than `postSecret` as `post` comes from the REST/HTTP layer. I left a [similar comment](https://github.com/elastic/elasticsearch/pull/103977#discussion_r1446246921) on one of Jedr's PRs, where I describe the rationale in more detail. Sorry for the nitpicking, but I think naming consistency can help us all in the future navigating the module :)

navarone-feekery replied:
It should be create, we don't support updating secrets

carlosdelest replied:
This solves it for me, thanks @navarone-feekery ! Agree on the naming nit, maybe `updateSecret` would be better. But, you know, _naming_...

navarone-feekery replied:
I changed it to `createSecret` in [this commit](https://github.com/elastic/elasticsearch/pull/103683/commits/9cf3f81a9e115dd26c15e24124d087a944ea4543).

navarone-feekery committed:
Fix typo and add FF check for index

navarone-feekery committed:
Move transport executions to index service class

navarone-feekery committed:
Re-introduce cluster privilege

navarone-feekery committed:
Add auth tests for endpoints

carlosdelest commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccounts.java:
```diff
@@ -49,6 +49,12 @@ final class ElasticServiceAccounts {
+                    .build(),
+                // Connector secrets. Enterprise Search can read and write to this index.
+                RoleDescriptor.IndicesPrivileges.builder()
+                    .indices(".connector-secrets*")
```
I think we no longer need explicit permissions for connector-secrets index, as that's taken care of by the OriginSettingClient + privileges combo

carlosdelest commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccountsTests.java:
```diff
@@ -379,6 +379,24 @@ public void testElasticEnterpriseSearchServerAccount() {
+        assertThat(role.cluster().check("cluster:admin/xpack/connector/secret/post", request, authentication), is(true));
+
+        final IndexAbstraction dotConnectorSecretsIndex = mockIndexAbstraction(".connector-secrets" + randomAlphaOfLengthBetween(1, 20));
+        assertThat(role.indices().allowedIndicesMatcher(TransportDeleteAction.NAME).test(dotConnectorSecretsIndex), is(true));
```
We don't need these checks, as this user won't be used to access the connector secrets, but the XPack internal user. We can add these tests [here](https://github.com/elastic/elasticsearch/blob/770671667eaad7eca8404785c0422bb54a4213f5/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/user/InternalUsersTests.java#L71).

carlosdelest commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/KibanaOwnedReservedRoleDescriptors.java:
```diff
@@ -378,7 +378,13 @@ static RoleDescriptor kibanaSystem(String name) {
                 RoleDescriptor.IndicesPrivileges.builder().indices(".logs-endpoint.heartbeat-*").privileges("read").build(),
                 // For connectors telemetry. Will be removed once we switched to connectors API
-                RoleDescriptor.IndicesPrivileges.builder().indices(".elastic-connectors*").privileges("read").build() },
+                RoleDescriptor.IndicesPrivileges.builder().indices(".elastic-connectors*").privileges("read").build(),
```
I guess we can use the privileges now instead of the raw index permissions, right?

carlosdelest commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:
```diff
@@ -1510,6 +1510,22 @@ public void testKibanaSystemRole() {
+
+        // Connector secrets. Kibana only has write access.
+        final IndexAbstraction dotConnectorSecretsIndex = mockIndexAbstraction(".connector-secrets");
+        assertThat(kibanaRole.indices().allowedIndicesMatcher("indices:foo").test(dotConnectorSecretsIndex), is(false));
```
Same here, we should check for privileges and not index permissions

navarone-feekery committed:
Remove outdated index permissions tests

navarone-feekery committed:
Add BWC and post/get unit tests

navarone-feekery committed:
Fix styling

navarone-feekery committed:
Add service index unit tests

navarone-feekery commented:
@timgrein I've added BWC and index service tests. Let me know if there are more tests I should add.
@carlosdelest I've removed the unnecessary index tests now that we rely on origin settings + privileges.

carlosdelest commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/KibanaOwnedReservedRoleDescriptors.java:
```diff
@@ -384,8 +384,12 @@ static RoleDescriptor kibanaSystem(String name) {
-                // For connectors telemetry. Will be removed once we switched to connectors API
-                RoleDescriptor.IndicesPrivileges.builder().indices(".elastic-connectors*").privileges("read").build() },
+                // Connector secrets. Kibana can only write to this index.
+                RoleDescriptor.IndicesPrivileges.builder()
```
Why do we need to use explicit index privileges vs using the new privileges for Kibana?

navarone-feekery replied:
We don't, I just missed this one 🤦🏻

navarone-feekery replied:
I've removed it

navarone-feekery committed:
Remove unnecessary kibana index privilege

navarone-feekery committed:
Re-add mistakenly removed code

navarone-feekery committed:
Merge branch 'main' into connector-secrets

carlosdelest reviewed (approved):
LGTM. Thanks for iterating on this! 🚀
